File: desktop/internal/db/projects.go
package db

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"time"
)

type Project struct {
	ProjectDir   string `json:"project_dir"`
	Name         string `json:"name"`
	Memories     int    `json:"memories"`
	UserMemories int    `json:"user_memories"`
	Issues       int    `json:"issues"`
	Tags         int    `json:"tags"`
}

func (d *DB) GetProjects() ([]Project, error) {
	projects := map[string]*Project{}

	// Memory counts per project
	rows, err := d.Query("SELECT project_dir, COUNT(*) as cnt FROM memories GROUP BY project_dir")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var pd string
		var cnt int
		if err := rows.Scan(&pd, &cnt); err != nil {
			return nil, err
		}
		if pd == "" {
			continue
		}
		projects[pd] = &Project{ProjectDir: pd, Memories: cnt}
	}

	// Issue counts (active + archive)
	for _, table := range []string{"issues", "issues_archive"} {
		rows, err := d.Query(fmt.Sprintf("SELECT project_dir, COUNT(*) as cnt FROM %s GROUP BY project_dir", table))
		if err != nil {
			return nil, err
		}
		for rows.Next() {
			var pd string
			var cnt int
			if err := rows.Scan(&pd, &cnt); err != nil {
				rows.Close()
				return nil, err
			}
			if pd == "" {
				continue
			}
			if _, ok := projects[pd]; !ok {
				projects[pd] = &Project{ProjectDir: pd}
			}
			projects[pd].Issues += cnt
		}
		rows.Close()
	}

	// Session state projects
	rows, err = d.Query("SELECT project_dir FROM session_state")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var pd string
		if err := rows.Scan(&pd); err != nil {
			return nil, err
		}
		if pd == "" {
			continue
		}
		if _, ok := projects[pd]; !ok {
			projects[pd] = &Project{ProjectDir: pd}
		}
	}

	// Tag counts per project
	tagRows, err := d.Query("SELECT project_dir, tags FROM memories")
	if err != nil {
		return nil, err
	}
	defer tagRows.Close()
	projTags := map[string]map[string]bool{}
	for tagRows.Next() {
		var pd, tagsJSON string
		if err := tagRows.Scan(&pd, &tagsJSON); err != nil {
			return nil, err
		}
		if pd == "" {
			continue
		}
		if _, ok := projTags[pd]; !ok {
			projTags[pd] = map[string]bool{}
		}
		var tags []string
		json.Unmarshal([]byte(tagsJSON), &tags)
		for _, t := range tags {
			projTags[pd][t] = true
		}
	}

	// User memory count & tags
	var userCount int
	if err := d.QueryRow("SELECT COUNT(*) FROM user_memories").Scan(&userCount); err != nil {
		log.Printf("scan error: %v", err)
	}

	userTags := map[string]bool{}
	uTagRows, err := d.Query("SELECT tags FROM user_memories")
	if err == nil {
		defer uTagRows.Close()
		for uTagRows.Next() {
			var tagsJSON string
			if err := uTagRows.Scan(&tagsJSON); err == nil {
				var tags []string
				json.Unmarshal([]byte(tagsJSON), &tags)
				for _, t := range tags {
					userTags[t] = true
				}
			}
		}
	}

	result := make([]Project, 0, len(projects))
	for pd, p := range projects {
		merged := map[string]bool{}
		for t := range projTags[pd] {
			merged[t] = true
		}
		for t := range userTags {
			merged[t] = true
		}
		parts := strings.Replace(pd, "\\", "/", -1)
		idx := strings.LastIndex(parts, "/")
		name := parts
		if idx >= 0 {
			name = parts[idx+1:]
		}
		p.Name = name
		p.UserMemories = userCount
		p.Tags = len(merged)
		result = append(result, *p)
	}
	return result, nil
}

func (d *DB) AddProject(projectDir string) error {
	projectDir = strings.TrimSpace(strings.Replace(projectDir, "\\", "/", -1))
	if projectDir == "" {
		return fmt.Errorf("project_dir is required")
	}
	now := time.Now().Format(time.RFC3339)
	_, err := d.Exec(
		"INSERT OR IGNORE INTO session_state (project_dir, is_blocked, block_reason, next_step, current_task, progress, recent_changes, pending, updated_at) VALUES (?,0,'','','','[]','[]','[]',?)",
		projectDir, now)
	return err
}

func (d *DB) DeleteProject(projectDir string) (int, error) {
	if projectDir == "" {
		return 0, fmt.Errorf("cannot delete empty project")
	}

	tx, err := d.Begin()
	if err != nil {
		return 0, err
	}
	defer d.UnlockAfterTx()

	// Get memory IDs
	rows, err := tx.Query("SELECT id FROM memories WHERE project_dir = ?", projectDir)
	if err != nil {
		tx.Rollback()
		return 0, err
	}
	var memIDs []string
	for rows.Next() {
		var id string
		rows.Scan(&id)
		memIDs = append(memIDs, id)
	}
	rows.Close()

	if len(memIDs) > 0 {
		ph := strings.Repeat("?,", len(memIDs))
		ph = ph[:len(ph)-1]
		args := make([]interface{}, len(memIDs))
		for i, id := range memIDs {
			args[i] = id
		}
		tx.Exec(fmt.Sprintf("DELETE FROM vec_memories WHERE id IN (%s)", ph), args...)
		tx.Exec(fmt.Sprintf("DELETE FROM memory_tags WHERE memory_id IN (%s)", ph), args...)
		tx.Exec(fmt.Sprintf("DELETE FROM memories WHERE id IN (%s)", ph), args...)
	}

	// Delete tasks
	tx.Exec("DELETE FROM tasks WHERE project_dir = ?", projectDir)
	tx.Exec("DELETE FROM tasks_archive WHERE project_dir = ?", projectDir)
	tx.Exec("DELETE FROM issues WHERE project_dir = ?", projectDir)
	tx.Exec("DELETE FROM issues_archive WHERE project_dir = ?", projectDir)
	tx.Exec("DELETE FROM session_state WHERE project_dir = ?", projectDir)

	if err := tx.Commit(); err != nil {
		return 0, err
	}
	return len(memIDs), nil
}

File: aivectormemory/web/routes/projects.py
from aivectormemory.db.memory_repo import MemoryRepo
from aivectormemory.db.user_memory_repo import UserMemoryRepo
from aivectormemory.db.issue_repo import IssueRepo


def get_stats(cm, pdir):
    repo = MemoryRepo(cm.conn, pdir)
    user_repo = UserMemoryRepo(cm.conn)
    issue_repo = IssueRepo(cm.conn, pdir)

    proj_count = repo.count(project_dir=pdir)
    user_count = user_repo.count()
    total_count = repo.count() + user_count

    all_issues, _ = issue_repo.list_by_date()
    status_counts = {}
    for i in all_issues:
        s = i["status"]
        status_counts[s] = status_counts.get(s, 0) + 1
    _, archived_total = issue_repo.list_archived()
    status_counts["archived"] = archived_total

    tag_counts = _merged_tag_counts(repo, user_repo, pdir)

    return {
        "memories": {"project": proj_count, "user": user_count, "total": total_count},
        "issues": status_counts,
        "tags": tag_counts,
    }


def _merged_tag_counts(mem_repo, user_repo, pdir):
    proj = mem_repo.get_tag_counts(project_dir=pdir)
    user = user_repo.get_tag_counts()
    merged = dict(proj)
    for k, v in user.items():
        merged[k] = merged.get(k, 0) + v
    return merged


def get_projects(cm):
    conn = cm.conn
    rows = conn.execute(
        "SELECT project_dir, COUNT(*) as mem_count FROM memories GROUP BY project_dir"
    ).fetchall()
    projects = {}
    for r in rows:
        pd = r["project_dir"]
        projects.setdefault(pd, {"project_dir": pd, "memories": 0, "issues": 0, "tag_count": 0})
        projects[pd]["memories"] = r["mem_count"]

    issue_rows = conn.execute(
        "SELECT project_dir, COUNT(*) as cnt FROM issues GROUP BY project_dir"
    ).fetchall()
    archive_rows = conn.execute(
        "SELECT project_dir, COUNT(*) as cnt FROM issues_archive GROUP BY project_dir"
    ).fetchall()
    for r in issue_rows:
        pd = r["project_dir"]
        projects.setdefault(pd, {"project_dir": pd, "memories": 0, "issues": 0, "tag_count": 0})
        projects[pd]["issues"] += r["cnt"]
    for r in archive_rows:
        pd = r["project_dir"]
        projects.setdefault(pd, {"project_dir": pd, "memories": 0, "issues": 0, "tag_count": 0})
        projects[pd]["issues"] += r["cnt"]

    state_rows = conn.execute("SELECT project_dir FROM session_state").fetchall()
    for r in state_rows:
        pd = r["project_dir"]
        projects.setdefault(pd, {"project_dir": pd, "memories": 0, "issues": 0, "tag_count": 0})

    tag_rows = conn.execute(
        "SELECT m.project_dir, COUNT(DISTINCT mt.tag) as tag_count "
        "FROM memories m JOIN memory_tags mt ON m.id = mt.memory_id "
        "GROUP BY m.project_dir"
    ).fetchall()
    for r in tag_rows:
        pd = r["project_dir"]
        if pd in projects:
            projects[pd]["tag_count"] = r["tag_count"]

    user_repo = UserMemoryRepo(conn)
    user_tag_count = conn.execute("SELECT COUNT(DISTINCT tag) FROM user_memory_tags").fetchone()[0]
    user_count = user_repo.count()

    result = []
    for pd, info in sorted(projects.items(), key=lambda x: -x[1]["memories"]):
        if not pd:
            continue
        result.append({
            "project_dir": pd,
            "name": pd.replace("\\", "/").rsplit("/", 1)[-1] if pd else "unknown",
            "memories": info["memories"],
            "user_memories": user_count,
            "issues": info["issues"],
            "tags": info.get("tag_count", 0) + user_tag_count,
        })
    return {"projects": result}


def add_project(handler, cm):
    from aivectormemory.web.api import _read_body
    body = _read_body(handler)
    project_dir = (body.get("project_dir") or "").strip()
    if not project_dir:
        return {"error": "project_dir is required"}
    project_dir = project_dir.replace("\\", "/")
    conn = cm.conn
    now = __import__("datetime").datetime.now().astimezone().isoformat()
    conn.execute(
        "INSERT OR IGNORE INTO session_state (project_dir, is_blocked, block_reason, next_step, current_task, progress, recent_changes, pending, updated_at) VALUES (?,0,'','','','[]','[]','[]',?)",
        (project_dir, now))
    conn.commit()
    return {"success": True, "project_dir": project_dir}


def browse_directory(params):
    import os
    path = (params.get("path", [None])[0] or "").strip()
    if not path:
        path = os.path.expanduser("~")
    path = os.path.expanduser(path)
    path = os.path.realpath(path)
    home = os.path.expanduser("~")
    if not path.startswith(home + os.sep) and path != home:
        return {"error": "access denied", "path": path}
    if not os.path.isdir(path):
        return {"error": "not a directory", "path": path}
    dirs = []
    try:
        for entry in sorted(os.scandir(path), key=lambda e: e.name.lower()):
            if entry.is_dir() and not entry.name.startswith("."):
                dirs.append(entry.name)
    except PermissionError:
        return {"error": "permission denied", "path": path}
    return {"path": path.replace("\\", "/"), "dirs": dirs}


def delete_project(cm, project_dir):
    if not project_dir:
        return {"success": False, "error": "Cannot delete empty project"}
    conn = cm.conn
    mem_ids = [r["id"] for r in conn.execute("SELECT id FROM memories WHERE project_dir = ?", (project_dir,)).fetchall()]
    if mem_ids:
        placeholders = ",".join("?" * len(mem_ids))
        conn.execute(f"DELETE FROM vec_memories WHERE id IN ({placeholders})", mem_ids)
        conn.execute(f"DELETE FROM memories WHERE id IN ({placeholders})", mem_ids)
    conn.execute("DELETE FROM issues WHERE project_dir = ?", (project_dir,))
    conn.execute("DELETE FROM issues_archive WHERE project_dir = ?", (project_dir,))
    conn.execute("DELETE FROM session_state WHERE project_dir = ?", (project_dir,))
    conn.commit()
    return {"success": True, "deleted_memories": len(mem_ids)}
